### ISSUE TITLE
CVE-2016-2107 OpenSSL vulnerabilitiy (critical)

### ISSUE THREAD

**Issue by gsantner** [Kurento/bugtracker#117]
**KMS Version**:
Version: 6.5.0
Found modules:
	Module: 'core' version '6.5.0'
	Module: 'elements' version '6.5.0'
	Module: 'filters' version '6.5.0'

**Other libraries versions**:
```
ii  gstreamer1.5-libav:amd64              1.7.0.1~20160530102916.35.g90e6470.trusty  amd64        libav plugin for GStreamer
ii  gstreamer1.5-nice:amd64               0.1.13.1~20160530102934.80.gd9ef50e.trusty amd64        ICE library (GStreamer plugin)
ii  gstreamer1.5-plugins-bad:amd64        1.8.0~20160530103612.645.gb68ba3e.trusty   amd64        GStreamer plugins from the "bad" set
ii  gstreamer1.5-plugins-base:amd64       1.8.0~20160530103212.387.g21728a3.trusty   amd64        GStreamer plugins from the "base" set
ii  gstreamer1.5-plugins-good:amd64       1.8.0~20160530103259.265.geca7cfe.trusty   amd64        GStreamer plugins from the "good" set
ii  gstreamer1.5-plugins-ugly:amd64       1.8.0~20160530103107.83.g4885bb3.trusty    amd64        GStreamer plugins from the "ugly" set
ii  gstreamer1.5-pulseaudio:amd64         1.8.0~20160530103259.265.geca7cfe.trusty   amd64        GStreamer plugin for PulseAudio
ii  gstreamer1.5-x:amd64                  1.8.0~20160530103212.387.g21728a3.trusty   amd64        GStreamer plugins for X11 and Pango
ii  kms-core-6.0                          6.5.0.20160530141614.trusty                amd64        Kurento core module
ii  kms-elements-6.0                      6.5.0.20160530154355.trusty                amd64        Kurento elements module
ii  kms-filters-6.0                       6.5.0.20160530155443.trusty                amd64        Kurento filters module
ii  kms-jsonrpc-1.0                       1.1.1.trusty                               amd64        Kurento jsonrpc library
ii  kmsjsoncpp                            1.6.3~20160530102631.40.gd78deb7.trusty    amd64        Kurento jsoncpp library
ii  kurento-media-server-6.0              6.5.0.20160530172436.trusty                amd64        Kurento Media Server
ii  libgstreamer-plugins-bad1.5-0:amd64   1.8.0~20160530103612.645.gb68ba3e.trusty   amd64        GStreamer development files for libraries from the "bad" set
ii  libgstreamer-plugins-base1.5-0:amd64  1.8.0~20160530103212.387.g21728a3.trusty   amd64        GStreamer libraries from the "base" set
ii  libgstreamer1.5-0:amd64               1.8.0~20160530103131.368.g0fb3a08.trusty   amd64        Core GStreamer libraries and elements
ii  libnice10:amd64                       0.1.13.1~20160530102934.80.gd9ef50e.trusty amd64        ICE library (shared library)
ii  libssl-dev:amd64                      1.0.2d-0ubuntu1                            amd64        Secure Sockets Layer toolkit - development files
ii  libssl-doc                            1.0.2d-0ubuntu1                            all          Secure Sockets Layer toolkit - development documentation
ii  libssl1.0.0:amd64                     1.0.2d-0ubuntu1                            amd64        Secure Sockets Layer toolkit - shared libraries
ii  openssl                               1.0.2d-0ubuntu1                            amd64        Secure Sockets Layer toolkit - cryptographic utility
ii  python-openssl                        0.13-2ubuntu6                              amd64        Python 2 wrapper around the OpenSSL library
ii  ssl-cert                              1.0.33                                     all          simple debconf wrapper for OpenSSL

```

**Client libraries**
- Language: Doesn't matter
- Version: Doesn't matter

**Browsers tested**
  Add OK or FAIL, along with the version, after browsers where you
have tested this issue:
- Chrome: Doesn't matter
- Firefox: Doesn't matter
- Native: Doesn't matter

**System description:**
Stun, Turn & KMS installed on ubuntu 14.04.

**What steps will reproduce the problem?**
1. Install Ubuntu 14.04
2. Install kms by repository

**What is the expected result?**
No security hole by kurento KMS

# What happens instead?
This issue is about the security hole in libssl1.0.0, named [CVE-2016-2107](https://blog.cloudflare.com/yet-another-padding-oracle-in-openssl-cbc-ciphersuites/).

Canonical/Ubuntu fixed this issue already with an update in the official repos of Ubuntu 14.04:
<http://www.ubuntu.com/usn/usn-2959-1/>  ->   libssl1.0.0 1.0.1f-1ubuntu2.19 

The big problem is Kurento requires libssl1.0.0 >= 1.0.2.
As Ubuntu doesn't have 1.0.2 in their repos Kurento choose to supply own packages of libssl in the own repository (`deb http://ubuntu.kurento.org trusty kms6`).
The supplied package is `pool/kms6/o/openssl/libssl1.0.0_1.0.2d-0ubuntu1_amd64.deb`

This means: It's not possible to install the fixed `1.0.1f-1ubuntu2.19` because this would break dependencies, and kurento creates an security hole by not (yet) supplying an security update.

Please update the package in your repository to **at least  1.0.2h**.

**Does it happen with one of the tutorials?**
YES
**Please provide any additional information below.**
apt-cache show  libssl1.0.0:
```
Package: libssl1.0.0
Version: 1.0.2d-0ubuntu1
Installed-Size: 3440
Priority: important
Section: libs
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Architecture: amd64
Description-en: Secure Sockets Layer toolkit - shared libraries
 This package is part of the OpenSSL project's implementation of the SSL
 and TLS cryptographic protocols for secure communication over the
 Internet.
 .
 It provides the libssl and libcrypto shared libraries.
Description-md5: 88547c6206c7fbc4fcc7d09ce100d210
MD5sum: 9e598a7638d15a1f3681298738aecede
SHA1: e705f861bf1d31671e607a89cf9d65699f2fb1dd
SHA256: 3d7c2925ba2d836c81593de70a590e5a01d8295f7e381cfb10a6e4110ab80852
Multi-Arch: same
Source: openssl
Filename: pool/kms6/o/openssl/libssl1.0.0_1.0.2d-0ubuntu1_amd64.deb
Depends: libc6 (>= 2.14), debconf (>= 0.5) | debconf-2.0
Size: 1091442
Original-Maintainer: Debian OpenSSL Team <pkg-openssl-devel@lists.alioth.debian.org>

```

Link to OpenSSL vulnerabilities page:
<https://www.openssl.org/news/vulnerabilities.html#toc>

* Fixed in OpenSSL 1.0.1t (Affected 1.0.1s, 1.0.1r, 1.0.1q, 1.0.1p, 1.0.1o, 1.0.1n, 1.0.1m, 1.0.1l, 1.0.1k, 1.0.1j, 1.0.1i, 1.0.1h, 1.0.1g, 1.0.1f, 1.0.1e, 1.0.1d, 1.0.1c, 1.0.1b, 1.0.1a, 1.0.1)
* Fixed in OpenSSL 1.0.2h (Affected 1.0.2g, 1.0.2f, 1.0.2e, 1.0.2d, 1.0.2c, 1.0.2b, 1.0.2a, 1.0.2)

I found this issue by checking the SSL Certificate.
Please fix this asap!

**Comment by jcaden** [comment]
Thanks for the report, we will try to fix it asap

**Comment by jcaden** [comment]
In the meanwhile, if you install 1.0.2f just downloading from xenial repository may do the trick. I haven't checked if this is possible because of the dependencies but it's probably.

**Comment by jcaden** [comment]
Just checked and installing xenial libssl version in trusty works.

**Comment by gsantner** [comment]
Thanks for your workaround, it works.
But note: The fix was backported to [1.0.2g on Ubuntu Xenial](http://changelogs.ubuntu.com/changelogs/pool/main/o/openssl/openssl_1.0.2g-1ubuntu4.1/changelog), not 1.0.2f
And OpenSSL still states their 1.0.2h is their first release with the fix included.

**Comment by jcaden** [comment]
Great, my idea was to provide 1.0.2g-1ubuntu4.1 so the fix should be included

**Comment by gsantner** [comment]
Are there any updates on this?

**Comment by jcaden** [comment]
It has been already updated, sorry for not notifying here.

**Comment by marcelrouw** [comment]
I just tested our ssl certificate in coturn using www.ssllabs.com. When I discovered OpenSSL Padding Oracle vulnerability (CVE-2016-2107) issue on my system. 

It seems this vulnerability isn't fixed in 1.0.2g. Probably it's fixed in de 1.0.2g Xenial version and not in the 1.0.2g Trusty version. But I'm no expert on this ;)

$ sudo apt-cache policy openssl
openssl:
  Installed: 1.0.2g-1ubuntu4
  Candidate: 1.0.2g-1ubuntu4
  Version table:
 *** 1.0.2g-1ubuntu4 0
        500 http://ubuntu.kurento.org/ trusty/kms6 amd64 Packages
        100 /var/lib/dpkg/status
     1.0.1f-1ubuntu2.21 0
        500 http://eu-west-1.ec2.archive.ubuntu.com/ubuntu/ trusty-updates/main amd64 Packages
        500 http://security.ubuntu.com/ubuntu/ trusty-security/main amd64 Packages
     1.0.1f-1ubuntu2 0
        500 http://eu-west-1.ec2.archive.ubuntu.com/ubuntu/ trusty/main amd64 Packages


I'm running the current version of kurento
Version: 6.6.1
Found modules:
	Module: 'core' version '6.6.1'
	Module: 'elements' version '6.6.1'
	Module: 'filters' version '6.6.1'

Kind Regards
Marcel
